### PR TITLE
refactor(threads): avoid unnecessary scheduler invocations

### PR DESCRIPTION
**Pull request by elenaf9**
# Description

This PR implements some changes and checks to avoid unnecessary scheduler invocations:
- the scheduler is now triggered in `set_state` as a central point
- if a thread changes into `ThreadState::Running`, only trigger the scheduler if the thread has a higher priority than the currently running one (**new**)
- always trigger the scheduler when a thread changes from `ThreadState::Running` into another state because this only ever happens if a currently running thread gets blocked
- only trigger the scheduler in `yield_now` if the runqueue contains any other threads

## Issues/PRs references

Extracted from #397.

## Benchmarks

(ticks **per iteration**)
`bench_sched_flags` is PR'd in #456

nRF52840dk: 

benchmark | main | this PR
| - | - | -
`bench_sched_yield` with two threads |  560 | 580
`bench_sched_yield` with one thread | 235 | 98
`bench_sched_flags` | 1848 |  1461

RPI-Pico

benchmark | main | this PR
| - | - | -
`bench_sched_yield` with two threads |  631 | 647
`bench_sched_yield` with one thread | 243 | 78
`bench_sched_flags` | 1739 |  1495

The additional checks come a some cost when the scheduler invocation is really needed, as visible in the `bench_sched_yield` benchmark. 
But I think this is tolerable, because the actual context switching logic (i.e. `sched`) is not affected, and in many cases we save an unnecessary scheduler call that would be much more expensive.

## Open Questions

See self review.

## Change checklist

- [x] I have cleaned up my commit history and squashed fixup commits.
- [x] I have followed the [Coding Conventions](https://future-proof-iot.github.io/RIOT-rs/dev/docs/book/coding-conventions.html).
- [x] I have performed a self-review of my own code.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: src/riot-rs-threads/src/lib.rs [hunk 1]
```diff
@@ -172,22 +168,29 @@ impl Threads {
...
+        if state == ThreadState::Running {
+            self.runqueue.add(pid, prio);
+            self.schedule_if_higher_prio(pid, prio);
+        } else if old_state == ThreadState::Running {
+            // A running thread is only set into a non-running state
+            // if it itself initiated it.
+            debug_assert_eq!(Some(pid), self.current_pid());
+
+            self.runqueue.pop_head(pid, prio);
+            schedule();
```
Comment (elenaf9) [inline src/riot-rs-threads/src/lib.rs:193]: I removed the checks that the respective other state is `!= ThreadState::Running`, because we never have that case in our code.

Interestingly, if I don't remove them `bench_sched_yield` only takes 574 instead of 580 ticks for the nRF52840, which is very weird given that the benchmark doesn't even use this method apart from the initial thread creation. Any idea why that could be @kaspar030?
(`bench_sched_flags` performs as expected better without the checks).

File: src/riot-rs-threads/src/lib.rs [hunk 2]
```diff
@@ -118,20 +118,16 @@ impl Threads {
...
-        } else {
-            None
-        }
+    ) -> Option<ThreadId> {
+        let (thread, pid) = self.get_unused()?;
+        Cpu::setup_stack(thread, stack, func, arg);
+        thread.prio = prio;
+        thread.pid = pid;
+        thread.state = ThreadState::Paused;
+        Some(pid)
```
Comment (elenaf9) [inline src/riot-rs-threads/src/lib.rs:127]: Drive-by improvement because I noticed that the caller doesn't need the `&mut Thread`.

File: src/riot-rs-threads/src/lib.rs [hunk 3]
```diff
@@ -202,37 +205,62 @@ impl Threads {
...
+
         if self.runqueue.peek_head(old_prio) == Some(thread_id) {
             self.runqueue.pop_head(thread_id, old_prio);
         } else {
             self.runqueue.del(thread_id);
         }
         self.runqueue.add(thread_id, prio);
-        true
+
+        // Check if the thread currently running and trigger the scheduler if
```
Comment (elenaf9) [inline src/riot-rs-threads/src/lib.rs:237]:
```suggestion
        // Check if the thread is currently running and trigger the scheduler if
```

File: src/riot-rs-threads/src/lib.rs [hunk 4]
```diff
@@ -202,37 +205,62 @@ impl Threads {
...
         let thread = self.get_unchecked_mut(thread_id);
         let old_prio = thread.prio;
         if old_prio == prio {
-            return false;
+            return;
         }
         thread.prio = prio;
         if thread.state != ThreadState::Running {
-            return false;
+            // No runqueue changes or scheduler invocations needed.
```
Comment (elenaf9) [inline src/riot-rs-threads/src/lib.rs:226]:
```suggestion
            // No runqueue changes or scheduler invocation needed.
```

File: src/riot-rs-threads/src/lib.rs [hunk 5]
```diff
@@ -172,22 +168,29 @@ impl Threads {
...
-            self.runqueue.pop_head(thread.pid, thread.prio);
+        let thread = self.get_unchecked_mut(pid);
+        let old_state = core::mem::replace(&mut thread.state, state);
+        let prio = thread.prio;
+
+        if state == ThreadState::Running {
+            self.runqueue.add(pid, prio);
+            self.schedule_if_higher_prio(pid, prio);
+        } else if old_state == ThreadState::Running {
+            // A running thread is only set into a non-running state
```
Comment (elenaf9) [inline src/riot-rs-threads/src/lib.rs:188]:
```suggestion
            // A running thread is only set to a non-running state
```

File: src/riot-rs-threads/src/lib.rs (hunk 1 above)
Comment (kaspar030) [reply src/riot-rs-threads/src/lib.rs]: > I removed the checks that the respective other state is `!= ThreadState::Running`, because we never have that case in our code.

Hm, while that may well be, are we ok if we made a mistake here? E.g., in the first case, `runqueue.add()` calls `runqueue.queues.push()`, which with a quick glance probably won't like if a thread is added a second time.

Comment (elenaf9) [reply src/riot-rs-threads/src/lib.rs]: `CList::push` prevents that thread is added twice, by early-returning if `self.next_idxs[n as usize] != Self::sentinel()`.

And in the second `else if` branch the check `state != ThreadState::Running` then is redundant because that's already catched by the first branch.

File: src/riot-rs-threads/src/lib.rs [hunk 6]
```diff
@@ -202,37 +205,62 @@ impl Threads {
...
+        }
+
+        // Thread isn't running.
+        // Only schedule if the thread has a higher priority than the running one.
+        if prio > old_prio {
+            self.schedule_if_higher_prio(thread_id, prio);
+        }
+    }
+
+    /// Triggers the scheduler if the thread has higher priority than the current running thread.
```
Comment (kaspar030) [inline src/riot-rs-threads/src/lib.rs:253]:
```suggestion
    /// Triggers the scheduler if the thread has higher priority than the currently running thread.
```

File: src/riot-rs-threads/src/lib.rs (hunk 1 above)
Comment (kaspar030) [reply src/riot-rs-threads/src/lib.rs]: no, fine, just verifying. :slightly_smiling_face:

File: src/riot-rs-threads/src/lib.rs (hunk 6 above)
Comment (elenaf9) [reply src/riot-rs-threads/src/lib.rs:253]:
```suggestion
    /// Triggers the scheduler if the thread has a higher priority than the currently running thread.
```

**Comment by elenaf9** [comment]
Rebased so that recently merged `bench_sched_flags` can be run on this branch.

**Comment by kaspar030** [comment]
LGTM!
